Sources/Client/Extensions/Functions.swift

```diff
@@ -0,0 +1,48 @@
+//
+//  Functions.swift
+//  StreamChatClient
+//
+//  Created by Alexey Bukhtin on 21/01/2020.
+//  Copyright © 2020 Stream.io Inc. All rights reserved.
+//
+
+// MARK: Side Effects
+
+/// A completion block type with a `Result`.
+public typealias Completion<T, E: Error> = (Result<T, E>) -> Void
+
+/// Performs side effect work for a success result before of the original completion block.
+/// - Parameters:
+///   - completion: an original completion block.
+///   - doBefore: a side effect will be executed before the original completion block.
+public func doBefore<T, E: Error>(_ completion: @escaping Completion<T, E>, _ doBefore: @escaping (T) -> Void) -> Completion<T, E> {
+    return addSideEffect(for: completion, doBefore: doBefore)
+}
+
+/// Performs side effect work for a success result after of the original completion block.
+/// - Parameters:
+///   - completion: an original completion block.
+///   - doAfter: a side effect will be executed after the original completion block.
+public func doAfter<T, E: Error>(_ completion: @escaping Completion<T, E>, _ doAfter: @escaping (T) -> Void) -> Completion<T, E> {
+    return addSideEffect(for: completion, doAfter: doAfter)
+}
+
+/// Performs side effect work for a success result before and after of the original completion block.
+/// - Parameters:
+///   - completion: an original completion block.
+///   - doBefore: a side effect will be executed before the original completion block.
+///   - doAfter: a side effect will be executed after the original completion block.
+public func addSideEffect<T, E: Error>(for completion: @escaping Completion<T, E>,
```

Comment:
Seems like a quick and dirty implementation for part of Promises! 😄 I suggest renaming `addSideEffect` to `do(_:before:after)` so:
```
addSideEffect(for: completion, doBefore: doBefore)
```
becomes:
```
do(completion, after: doBefore)
```
and
```
addSideEffect(for: completion, doAfter: doAfter)
```
becomes:
```
do(completion, before: doAfter)
```

Obviously this is counter-intuitive (`after` and `before` parameters) so another suggestion is rename `addSideEffect` to one of these:
* `organize`
* `order`
* `doInOrder`

Reply:
It was `do`, but I thought it's not clear and renamed. I'll rename it back.